### PR TITLE
rgw: remove redundant codes in rgw_cache.h

### PR DESCRIPTION
**Pull request by Wilhelmshaven**
I'm not sure whether the old way is faster or not, seems unnecessary to do that with unsafe sprintf.

Signed-off-by: lihongjie <lihongjie@cmss.chinamobile.com>

### CODE REVIEW

**Comment by liewegas** [comment]
rebase please?

**Comment by Wilhelmshaven** [comment]
@liewegas rebased, tkx

**Comment by cbodley** [comment]
> I'm not sure whether the old way is faster or not, seems unnecessary to do that with unsafe sprintf.

i agree that we shouldn't be using `sprintf()`, thanks. but this pattern of `std::string::append()` could lead to extra allocations. could you add some math to calculate the resulting string length, and use `std::string::reserve()` to make sure we only allocate once? something like this, maybe:
```c++
std::string buf;
buf.reserve(pool.name.size() + oid.size() + 2);
return buf.append(pool.name).append('+').append(oid);
```

**Comment by Wilhelmshaven** [comment]
@cbodley Thanks for your really helpful advise :)

File: src/rgw/rgw_cache.h [hunk 1]
```diff
@@ -181,12 +181,9 @@ class RGWCache  : public T
...
   string normal_name(rgw_pool& pool, const std::string& oid) {
-    string& pool_name = pool.name;
-    char buf[pool_name.size() + 1 + oid.size() + 1];
-    const char *pool_str = pool_name.c_str();
-    const char *oid_str = oid.c_str();
-    sprintf(buf, "%s+%s", pool_str, oid_str);
-    return string(buf);
+    std::string buf;
+    buf.reserve(pool.name.size() + oid.size() + 2);
+    return buf.append(pool.name).append("+").append(oid);
```
Comment (cbodley) [inline src/rgw/rgw_cache.h]: oops sorry, this last line will interfere with [return value optimization](https://en.wikipedia.org/wiki/Return_value_optimization), and induce an extra alloc/copy back to the caller. splitting up the `append()`s and `return` will help:
```c++
    buf.append(pool.name).append("+").append(oid);
    return buf;
```

Comment (Wilhelmshaven) [inline src/rgw/rgw_cache.h]: Oh, okay, fixed, thx. :)

**Comment by cbodley** [comment]
passed teuthology: http://pulpito.ceph.com/cbodley-2017-03-21_11:46:22-rgw-wip-cbodley-testing---basic-mira/

thanks!

**Pull request merged**